Fix slow compilation of AddressSectionElement.updateAddressFields

matthewc13 opened the pull request:
## Summary
<!-- Simple summary of what was changed. -->

AddressSectionElement.updateAddressFields is currently extremely slow to compile taking around ~8004.12ms to compile. This is due to array concatenation of different derived types. By providing swift with base class type information for elements this drastically speeds up compilation resulting in compilation for this function now taking ~14.67ms. 

## Motivation
<!-- Why are you making this change? If it's for fixing a bug, if possible, please include a code snippet or example project that demonstrates the issue. -->

Decrease compilation time of the Stripe package, gaining 7ish seconds of our CI time back per run. 

## Testing
<!-- How was the code tested? Be as specific as possible. -->
To validate the difference run the following on the master branch and this branch respectively. 

Perform a clean build from terminal using the following: 
`xcodebuild -workspace Stripe.xcworkspace  -scheme "PaymentSheet Example" clean build -xcconfig Example/PaymentSheet\ Example/BuildConfigurations/PaymentSheet-Example-Debug.xcconfig -destination "platform=iOS Simulator,name=iPhone 8,OS=15.2"  -derivedDataPath /tmp/ddata OTHER_SWIFT_FLAGS="-Xfrontend -debug-time-function-bodies" | grep ".[0-9]ms" | grep -v "^0.[0-9]ms" | sort -nr > ~/Desktop/timer.txt`

`open ~/Desktop/timer.txt`

Search for `updateAddressFields` (on master this will be the first entry)

compare 

## Changelog
<!-- Is this a notable change that affects users? If so, add a line to `CHANGELOG.md` and prefix the line with one of the following:
    - [Added] for new features.
    - [Changed] for changes in existing functionality.
    - [Deprecated] for soon-to-be removed features.
    - [Removed] for now removed features.
    - [Fixed] for any bug fixes.
    - [Security] in case of vulnerabilities.
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/stripe/stripe-ios?pullRequest=2239) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/stripe/stripe-ios?pullRequest=2239) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/stripe/stripe-ios?pullRequest=2239) it.</sub>

eurias-stripe commented on README.md:
```diff
@@ -141,6 +141,12 @@ You can demo this feature in our [PaymentSheet example app](Example/PaymentSheet
 
 We welcome contributions of any kind including new features, bug fixes, and documentation improvements. Please first open an issue describing what you want to build if it is a major change so that we can discuss how to move forward. Otherwise, go ahead and open a pull request for minor changes such as typo fixes and one liners.
 
+### Generating workspace
+
+We use Tuist to generate a workspace. Please install Tuise following the [Get started guide](https://docs.tuist.io/tutorial/get-started)
+
+After installation run `tuist generate` to generate the Stripe workspace.
+
```
These instructions are already included in the Building from source section, so they are redundant, please remove.

matthewc13 replied:
Sorry must have missed this

eurias-stripe commented on StripeUICore/StripeUICore/Source/Elements/Factories/Address/AddressSectionElement.swift:
```diff
@@ -334,8 +334,10 @@ import UIKit
+        
+        let initialElements: [Element?] = [name, country, autoCompleteLine]
+        let phoneElement: [Element?] = [phone]
+        addressSection.elements = (initialElements + addressFields + phoneElement).compactMap { $0 }
```
This `compactMap` call shouldn't be needed anymore (I don't know if it was needed in the first place), the arrays should be able to be concatenated directly.

matthewc13 replied:
name, country etc can all be nil, they are properties on the class.

eurias-stripe replied:
Duh! For some reason I didn't realize they were optionals and thought `compactMap` was used to cast into the right type. My bad.

eurias-stripe reviewed (changes requested):
Wow, this is a massive win! Thanks for the contribution @matthewc13.
I left a couple of comments, and it looks like CI is failing the lint check, but overall it looks great.

matthewc13 commented:
@eurias-stripe Hopefully addressed all your comments and the linting issue. Is there a way to locally run your linter that I am missing? Hopefully this should be enough to get this going anyways :)

eurias-stripe commented:
Thank you for your contribution!
